File: service/apiclient/internal/restclient/token.go
package restclient

import (
	"context"
	"fmt"
	"net/url"
)

type TokenService service

type TokenResponse struct {
    Data Token `json:"data"`
}

type Token struct {
	Token          *string        `json:"token,omitempty"`
	AuthID         *string        `json:"auth_id,omitempty"`
	AccentUserUUID *string        `json:"accent_user_uuid,omitempty"`
	AccentUUID     *string        `json:"accent_uuid,omitempty"`
	IssuedAt       *string        `json:"issued_at,omitempty"`
	ExpiresAt      *string        `json:"expires_at,omitempty"`
	UTCIssuedAt    *string        `json:"utc_issued_at,omitempty"`
	UTCExpiresAt   *string        `json:"utc_expires_at,omitempty"`
	ACL            []*string      `json:"acl,omitempty"`
	SessionUUID    *string        `json:"session_uuid,omitempty"`
	RemoteAddr     *string        `json:"remote_addr,omitempty"`
	UserAgent      *string        `json:"user_agent,omitempty"`
	Metadata       *TokenMetadata `json:"metadata,omitempty"`
}

type TokenMetadata struct {
	UUID        *string `json:"uuid,omitempty"`
	TenantUUID  *string `json:"tenant_uuid,omitempty"`
	AuthID      *string `json:"auth_id,omitempty"`
	PBXUserUUID *string `json:"pbx_user_uuid,omitempty"`
	AccentUUID  *string `json:"accent_uuid,omitempty"`
	Purpose     *string `json:"purpose,omitempty"`
	Admin       *bool   `json:"admin,omitempty"`
}

// TokenRequestParams specifies the parameters for creating a new token.
type TokenRequestParams struct {
	AccessType    string `json:"access_type,omitempty"`
	Backend       string `json:"backend,omitempty"`
	ClientID      string `json:"client_id,omitempty"`
	DomainName    string `json:"domain_name,omitempty"`
	Expiration    int    `json:"expiration,omitempty"`
	RefreshToken  string `json:"refresh_token,omitempty"`
	Username      string `json:"username,omitempty"`
	Password      string `json:"password,omitempty"`
}

// func (t TokenResponse) String() string {
// 	return Stringify(t)
// }

func (t Token) String() string {
	return Stringify(t)
}

/*
  /token:
    post:
      consumes:
      - application/json
      description: 'Creates a valid token for the supplied username and password combination
        or refresh_token

        using the specified backend.


        The user''s email address can be used instead of the username if the email
        address is confirmed.


        The stock backends are: ``accent_user``, ``ldap_user``.


        Creating  a token with the `access_type` *offline* will also create a refresh
        token which can be used

        to create a new token without specifying the username and password.


        The username/password and refresh_token method of authentication are mutually
        exclusive


        For more details about the backends, see http://documentation.accent.community/en/latest/system/accent-auth/stock_plugins.html#backends-plugins

        '
      operationId: createToken
      parameters:
      - $ref: '#/parameters/accent_session_type'
      - description: The token creation parameters
        in: body
        name: body
        required: false
        schema:
          properties:
            access_type:
              default: online
              description: 'The `access_type` indicates whether your application can
                refresh the tokens when the user is not

                present at the browser. Valid parameter values are *online*, which
                is the default value, and *offline*


                Only one refresh token will be created for a given user with a given
                `client_id`. The old refresh

                for `client_id` will be revoken when creating a new one.


                The *client_id* field is required when using the `access_type` *offline*

                '
              enum:
              - online
              - offline
              type: string
            backend:
              default: accent_user
              type: string
            client_id:
              description: 'The `client_id` is used in conjunction with the `access_type`
                *offline* to known for which application

                a refresh token has been emitted.


                *Required when using `access_type: offline`*

                '
              type: string
            domain_name:
              description: 'The `domain_name` must match a tenant''s domain_name entry
                to find the appropriate ldap configuration.

                '
              type: string
            expiration:
              default: 7200
              description: Expiration time in seconds.
              maximum: 315360000
              type: integer
            refresh_token:
              description: 'The `refresh_token` can be used to get a new access token
                without using the username/password.

                This is useful for client application that should not store the username
                and password once the

                user has logged in a first time.

                '
              type: string
          type: object
      produces:
      - application/json
      responses:
        '200':
          description: The created token's data
          schema:
            $ref: '#/definitions/Token'
        '400':
          description: Invalid expiration or missing field
          schema:
            $ref: '#/definitions/Error'
        '500':
          description: System related token generation error
          schema:
            $ref: '#/definitions/Error'
      security:
      - accent_auth_basic: []
      summary: Creates a token
      tags:
      - token

			accent_session_type:
    description: The session type
    enum:
    - mobile
    - desktop
    in: header
    name: Accent-Session-Type
    required: false
    type: string


		Token:
    properties:
      data:
        properties:
          acl:
            description: The list of allowed accesses for this token
            items:
              type: string
            type: array
          auth_id:
            description: The unique identifier retrieved from the backend
            type: string
          expires_at:
            type: string
          issued_at:
            type: string
          metadata:
            description: Information owned by accent-auth about this user
            type: object
          session_uuid:
            type: string
          token:
            type: string
          utc_expires_at:
            type: string
          utc_issued_at:
            type: string
          accent_user_uuid:
            description: 'The UUID of the matching accent-confd user if there is one.
              This

              field can be null.


              This field should NOT be used anymore, the "pbx_user_uuid" in the

              metadata field is the prefered method to access this information.

              '
            type: string
          accent_uuid:
            type: string
        type: object
    type: object

			*/


// CreateToken creates a new token based on the provided parameters.
func (s *TokenService) CreateToken(ctx context.Context, params *TokenRequestParams) (*Token, *Response, error) {
	// Validate mutual exclusivity of Username/Password and RefreshToken
    if (params.Username != "" || params.Password != "") && params.RefreshToken != "" {
        return nil, nil, fmt.Errorf("username/password and refresh_token methods of authentication are mutually exclusive")
    }

    // If using Username and Password, ensure both are provided
    if (params.Username != "" || params.Password != "") && (params.Username == "" || params.Password == "") {
        return nil, nil, fmt.Errorf("both username and password must be provided for basic authentication")
    }

	prefix_path := "api/auth/0.1/"
	suffix_path := "token"
	u := fmt.Sprintf("%s%s", prefix_path, suffix_path)
	// Since this is a POST request, the body should be included.
	// Note: The API might require setting specific headers or other authentication methods.
	req, err := s.client.NewRequest("POST", u, params)
	if err != nil {
		return nil, nil, err
	}

	// Set any necessary headers here. For example, if session type is required:
	// req.Header.Set("Accent-Session-Type", "desktop")
	if params.Username != "" && params.Password != "" {
		req.SetBasicAuth(params.Username, params.Password)
  }

	tokenResp := new(TokenResponse)
	resp, err := s.client.Do(ctx, req, tokenResp)
	if err != nil {
		return nil, resp, err
	}

	return &tokenResp.Data, resp, nil
}



/*
  /token/{token}:
    delete:
      parameters:
      - $ref: '#/parameters/token'
      responses:
        '200':
          description: Success message
        '500':
          description: System related token error
          schema:
            $ref: '#/definitions/Error'
      security:
      - {}
      summary: Revoke a token
      tags:
      - token
    get:
      description: Checks if a token is valid in a given context and return the token
        data.  If a scope is given, the token must have the necessary permissions
        for the ACL. If a tenant is given, the token must have that tenant in its
        sub-tenant subtree.
      parameters:
      - $ref: '#/parameters/token'
      - $ref: '#/parameters/scope'
      - $ref: '#/parameters/tenant'
      responses:
        '200':
          description: The token's data
          schema:
            $ref: '#/definitions/Token'
        '403':
          description: This token cannot acces the required ACL
          schema:
            $ref: '#/definitions/Error'
        '404':
          description: Token not found
          schema:
            $ref: '#/definitions/Error'
        '500':
          description: System related token error
          schema:
            $ref: '#/definitions/Error'
      security:
      - {}
      summary: Retrieves token data
      tags:
      - token
    head:
      description: Checks if a token is valid in a given context.  If a scope is given,
        the token must have the necessary permissions for the ACL. If a tenant is
        given, the token must have that tenant in its sub-tenant subtree.
      parameters:
      - $ref: '#/parameters/token'
      - $ref: '#/parameters/scope'
      - $ref: '#/parameters/tenant'
      responses:
        '204':
          description: No data
        '403':
          description: This token cannot acces the required ACL
          schema:
            $ref: '#/definitions/Error'
        '404':
          description: Token not found
          schema:
            $ref: '#/definitions/Error'
        '500':
          description: System related token error
          schema:
            $ref: '#/definitions/Error'
      security:
      - {}
      summary: Checks if a token is valid
      tags:
      - token

  token:
    description: The token to query
    in: path
    name: token
    required: true
    type: string

  scope:
    description: The required ACL
    in: query
    name: scope
    required: false
    type: string

  tenant:
    description: A tenant UUID to check against
    in: query
    name: tenant
    required: false
    type: string

	  Error:
    properties:
      reason:
        items:
          type: string
        type: array
      status_code:
        type: integer
      timestamp:
        items:
          type: string
        type: array
    type: object

		Token:
    properties:
      data:
        properties:
          acl:
            description: The list of allowed accesses for this token
            items:
              type: string
            type: array
          auth_id:
            description: The unique identifier retrieved from the backend
            type: string
          expires_at:
            type: string
          issued_at:
            type: string
          metadata:
            description: Information owned by accent-auth about this user
            type: object
          session_uuid:
            type: string
          token:
            type: string
          utc_expires_at:
            type: string
          utc_issued_at:
            type: string
          accent_user_uuid:
            description: 'The UUID of the matching accent-confd user if there is one.
              This

              field can be null.


              This field should NOT be used anymore, the "pbx_user_uuid" in the

              metadata field is the prefered method to access this information.

              '
            type: string
          accent_uuid:
            type: string
        type: object
    type: object
*/

// RevokeToken revokes an existing token.
func (s *TokenService) RevokeToken(ctx context.Context, token string) (*Response, error) {
	u := fmt.Sprintf("token/%s", token)
	req, err := s.client.NewRequest("DELETE", u, nil)
	if err != nil {
		return nil, err
	}

	return s.client.Do(ctx, req, nil)
}

// GetToken retrieves token data, checking its validity in a given context.
func (s *TokenService) GetToken(ctx context.Context, token, scope, tenant string) (*Token, *Response, error) {
	u := fmt.Sprintf("token/%s", token)
	// Add query parameters for scope and tenant if provided
	queryParams := url.Values{}
	if scope != "" {
		queryParams.Add("scope", scope)
	}
	if tenant != "" {
		queryParams.Add("tenant", tenant)
	}
	if len(queryParams) > 0 {
		u = fmt.Sprintf("%s?%s", u, queryParams.Encode())
	}

	req, err := s.client.NewRequest("GET", u, nil)
	if err != nil {
		return nil, nil, err
	}

	tokenResp := new(TokenResponse)
	resp, err := s.client.Do(ctx, req, tokenResp)
	if err != nil {
		return nil, resp, err
	}

	return &tokenResp.Data, resp, nil
}

// CheckTokenValidity checks if a token is valid in a given context.
func (s *TokenService) CheckTokenValidity(ctx context.Context, token, scope, tenant string) (*Response, error) {
	u := fmt.Sprintf("token/%s", token)
	// Add query parameters for scope and tenant if provided
	queryParams := url.Values{}
	if scope != "" {
		queryParams.Add("scope", scope)
	}
	if tenant != "" {
		queryParams.Add("tenant", tenant)
	}
	if len(queryParams) > 0 {
		u = fmt.Sprintf("%s?%s", u, queryParams.Encode())
	}

	req, err := s.client.NewRequest("HEAD", u, nil)
	if err != nil {
		return nil, err
	}

	return s.client.Do(ctx, req, nil)
}

File: library/accent-auth-client/accent_auth_client/commands/token.py
# Copyright 2025 Accent Communications

from __future__ import annotations

import asyncio
import logging
from functools import lru_cache
from typing import Any, cast

import httpx
from accent_lib_rest_client import RESTCommand

from accent_auth_client.exceptions import (
    InvalidTokenException,
    MissingPermissionsTokenException,
)
from accent_auth_client.type_definitions import JSON, TokenDict

logger = logging.getLogger(__name__)


class TokenCommand(RESTCommand):
    """Command for token-related operations.

    Provides methods for creating, validating, and managing authentication tokens.
    """

    resource = "token"
    _user_agent = "Accent Python auth client"

    async def new_async(
        self,
        backend: str | None = None,
        expiration: int | None = None,
        session_type: str | None = None,
        user_agent: str | None = None,
        access_type: str | None = None,
        client_id: str | None = None,
        refresh_token: str | None = None,
        username: str | None = None,
        password: str | None = None,
        tenant_id: str | None = None,
        domain_name: str | None = None,
    ) -> TokenDict:
        """Create a new token asynchronously.

        Args:
            backend: Authentication backend name
            expiration: Token expiration time in seconds
            session_type: Type of session
            user_agent: User agent string
            access_type: Type of access
            client_id: Client identifier
            refresh_token: Refresh token
            username: Username for authentication
            password: Password for authentication
            tenant_id: Tenant identifier
            domain_name: Domain name

        Returns:
            TokenDict: Created token information

        Raises:
            AccentAPIError: If the request fails

        """
        data: dict[str, Any] = {}
        if backend:
            data["backend"] = backend
        if expiration:
            data["expiration"] = expiration
        if client_id:
            data["client_id"] = client_id
        if refresh_token:
            data["refresh_token"] = refresh_token
        if access_type:
            data["access_type"] = access_type
        if tenant_id:
            data["tenant_id"] = tenant_id
        if domain_name:
            data["domain_name"] = domain_name

        headers = self._get_headers()
        headers["User-Agent"] = self._user_agent
        if session_type:
            headers["Accent-Session-Type"] = session_type
        if user_agent:
            headers["User-Agent"] = user_agent

        auth = None
        if username and password:
            auth = httpx.BasicAuth(username=username, password=password)

        client = self.async_client
        if auth:
            client.auth = auth

        start_time = asyncio.get_event_loop().time()

        r = await client.post(self.base_url, headers=headers, json=data)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError as e:
            logger.error(
                "Failed to create token: %s, status code: %s", str(e), r.status_code
            )
            self.raise_from_response(r)

        response = self.process_json_response(r, start_time)
        return cast(TokenDict, response.data["data"])

    def new(
        self,
        backend: str | None = None,
        expiration: int | None = None,
        session_type: str | None = None,
        user_agent: str | None = None,
        access_type: str | None = None,
        client_id: str | None = None,
        refresh_token: str | None = None,
        username: str | None = None,
        password: str | None = None,
        tenant_id: str | None = None,
        domain_name: str | None = None,
    ) -> TokenDict:
        """Create a new token.

        Args:
            backend: Authentication backend name
            expiration: Token expiration time in seconds
            session_type: Type of session
            user_agent: User agent string
            access_type: Type of access
            client_id: Client identifier
            refresh_token: Refresh token
            username: Username for authentication
            password: Password for authentication
            tenant_id: Tenant identifier
            domain_name: Domain name

        Returns:
            TokenDict: Created token information

        Raises:
            AccentAPIError: If the request fails

        """
        data: dict[str, Any] = {}
        if backend:
            data["backend"] = backend
        if expiration:
            data["expiration"] = expiration
        if client_id:
            data["client_id"] = client_id
        if refresh_token:
            data["refresh_token"] = refresh_token
        if access_type:
            data["access_type"] = access_type
        if tenant_id:
            data["tenant_id"] = tenant_id
        if domain_name:
            data["domain_name"] = domain_name

        headers = self._get_headers()
        headers["User-Agent"] = self._user_agent
        if session_type:
            headers["Accent-Session-Type"] = session_type
        if user_agent:
            headers["User-Agent"] = user_agent

        auth = None
        if username and password:
            auth = httpx.BasicAuth(username=username, password=password)

        client = self.sync_client
        if auth:
            client.auth = auth

        r = client.post(self.base_url, headers=headers, json=data)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError:
            self.raise_from_response(r)

        response = self.process_json_response(r)
        return cast(TokenDict, response.data["data"])

    async def delete_async(
        self, user_uuid: str, client_id: str, tenant_uuid: str | None = None
    ) -> None:
        """Delete a token asynchronously.

        Args:
            user_uuid: User identifier
            client_id: Client identifier
            tenant_uuid: Optional tenant identifier

        Raises:
            AccentAPIError: If the request fails

        """
        headers = self._get_headers(tenant_uuid=tenant_uuid)
        url = self._client.url("users", user_uuid, "tokens", client_id)

        r = await self.async_client.delete(url, headers=headers)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError as e:
            if r.status_code != 204:
                logger.error(
                    "Failed to delete token: %s, status code: %s", str(e), r.status_code
                )
                self.raise_from_response(r)

    def delete(
        self, user_uuid: str, client_id: str, tenant_uuid: str | None = None
    ) -> None:
        """Delete a token.

        Args:
            user_uuid: User identifier
            client_id: Client identifier
            tenant_uuid: Optional tenant identifier

        Raises:
            AccentAPIError: If the request fails

        """
        headers = self._get_headers(tenant_uuid=tenant_uuid)
        url = self._client.url("users", user_uuid, "tokens", client_id)

        r = self.sync_client.delete(url, headers=headers)

        if r.status_code != 204:
            self.raise_from_response(r)

    async def revoke_async(self, token: str) -> None:
        """Revoke a token asynchronously.

        Args:
            token: Token to revoke

        Raises:
            AccentAPIError: If the request fails

        """
        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        await self.async_client.delete(url, headers=headers)

    def revoke(self, token: str) -> None:
        """Revoke a token.

        Args:
            token: Token to revoke

        Raises:
            AccentAPIError: If the request fails

        """
        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        self.sync_client.delete(url, headers=headers)

    async def check_async(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> bool:
        """Check if a token is valid asynchronously.

        Args:
            token: Token to check
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            bool: True if the token is valid

        Raises:
            InvalidTokenException: If the token is invalid
            MissingPermissionsTokenException: If the token lacks permissions
            AccentAPIError: For other API errors

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = await self.async_client.head(url, headers=headers, params=params)

        if r.status_code == 204:
            return True
        if r.status_code == 404:
            raise InvalidTokenException()
        if r.status_code == 403:
            raise MissingPermissionsTokenException()
        self.raise_from_response(r)
        return False

    def check(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> bool:
        """Check if a token is valid.

        Args:
            token: Token to check
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            bool: True if the token is valid

        Raises:
            InvalidTokenException: If the token is invalid
            MissingPermissionsTokenException: If the token lacks permissions
            AccentAPIError: For other API errors

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = self.sync_client.head(url, headers=headers, params=params)

        if r.status_code == 204:
            return True
        if r.status_code == 404:
            raise InvalidTokenException()
        if r.status_code == 403:
            raise MissingPermissionsTokenException()
        self.raise_from_response(r)
        return False

    @lru_cache(maxsize=128)
    async def is_valid_async(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> bool:
        """Check if a token is valid without raising exceptions (async).

        Args:
            token: Token to check
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            bool: True if the token is valid

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = await self.async_client.head(url, headers=headers, params=params)

        if r.status_code in (204, 403, 404):
            return r.status_code == 204
        try:
            self.raise_from_response(r)
        except Exception as e:
            logger.exception("Error checking token validity: %s", str(e))
            return False
        return False

    @lru_cache(maxsize=128)
    def is_valid(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> bool:
        """Check if a token is valid without raising exceptions.

        Args:
            token: Token to check
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            bool: True if the token is valid

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = self.sync_client.head(url, headers=headers, params=params)

        if r.status_code in (204, 403, 404):
            return r.status_code == 204
        try:
            self.raise_from_response(r)
        except Exception as e:
            logger.exception("Error checking token validity: %s", str(e))
            return False
        return False

    async def check_scopes_async(
        self, token: str, scopes: list[str], tenant: str | None = None
    ) -> JSON:
        """Check token scopes asynchronously.

        Args:
            token: Token to check
            scopes: List of scopes to check
            tenant: Optional tenant identifier

        Returns:
            JSON: Response data

        Raises:
            AccentAPIError: If the request fails

        """
        data: dict[str, Any] = {"scopes": scopes}
        if tenant:
            data["tenant_uuid"] = tenant

        headers = self._get_headers()
        headers["User-Agent"] = self._user_agent
        url = f"{self.base_url}/{token}/scopes/check"

        r = await self.async_client.post(url, headers=headers, json=data)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError:
            self.raise_from_response(r)

        return cast(JSON, r.json())

    def check_scopes(
        self, token: str, scopes: list[str], tenant: str | None = None
    ) -> JSON:
        """Check token scopes.

        Args:
            token: Token to check
            scopes: List of scopes to check
            tenant: Optional tenant identifier

        Returns:
            JSON: Response data

        Raises:
            AccentAPIError: If the request fails

        """
        data: dict[str, Any] = {"scopes": scopes}
        if tenant:
            data["tenant_uuid"] = tenant

        headers = self._get_headers()
        headers["User-Agent"] = self._user_agent
        url = f"{self.base_url}/{token}/scopes/check"

        r = self.sync_client.post(url, headers=headers, json=data)

        if r.status_code != 200:
            self.raise_from_response(r)

        return cast(JSON, r.json())

    async def get_async(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> TokenDict:
        """Get token information asynchronously.

        Args:
            token: Token to get
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            TokenDict: Token information

        Raises:
            AccentAPIError: If the request fails

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = await self.async_client.get(url, headers=headers, params=params)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError:
            self.raise_from_response(r)

        return cast(TokenDict, r.json()["data"])

    def get(
        self, token: str, required_acl: str | None = None, tenant: str | None = None
    ) -> TokenDict:
        """Get token information.

        Args:
            token: Token to check
            required_acl: Optional required ACL
            tenant: Optional tenant identifier

        Returns:
            TokenDict: Token information

        Raises:
            AccentAPIError: If the request fails

        """
        params = {}
        if required_acl:
            params["scope"] = required_acl
        if tenant:
            params["tenant"] = tenant

        headers = self._get_headers()
        url = f"{self.base_url}/{token}"

        r = self.sync_client.get(url, headers=headers, params=params)

        if r.status_code != 200:
            self.raise_from_response(r)

        return cast(TokenDict, r.json()["data"])

    async def list_async(
        self, user_uuid: str | None = None, **kwargs: Any
    ) -> list[TokenDict]:
        """List tokens for a user asynchronously.

        Args:
            user_uuid: User identifier (required)
            **kwargs: Additional parameters

        Returns:
            list[TokenDict]: List of tokens

        Raises:
            TypeError: If user_uuid is None
            AccentAPIError: If the request fails

        """
        if user_uuid is None:
            raise TypeError("user_uuid cannot be None")

        headers = self._get_headers(**kwargs)
        url = self._client.url("users", user_uuid, "tokens")

        r = await self.async_client.get(url, headers=headers, params=kwargs)

        try:
            r.raise_for_status()
        except httpx.HTTPStatusError:
            self.raise_from_response(r)

        return cast(list[TokenDict], r.json())

    def list(self, user_uuid: str | None = None, **kwargs: Any) -> list[TokenDict]:
        """List tokens for a user.

        Args:
            user_uuid: User identifier (required)
            **kwargs: Additional parameters

        Returns:
            list[TokenDict]: List of tokens

        Raises:
            TypeError: If user_uuid is None
            AccentAPIError: If the request fails

        """
        if user_uuid is None:
            raise TypeError("user_uuid cannot be None")

        headers = self._get_headers(**kwargs)
        url = self._client.url("users", user_uuid, "tokens")

        r = self.sync_client.get(url, headers=headers, params=kwargs)

        if r.status_code != 200:
            self.raise_from_response(r)

        return cast(list[TokenDict], r.json())
